File: C/Starting/04_gather_parallel_dart.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>
#include <mpi.h>
#define ITS 10000

void seed_rng(int rank){
  time_t t;
  srand((unsigned) time(&t) + rank);
}

double get_random_in_range(){
  return rand()%100000/99999.0 * 2.0 - 1.0;
}

int main(int argc, char **argv){
  int count = 0, index;
  double d1, d2;
  int rank, nproc;
  int *count_global;
  MPI_Init(&argc, &argv);
  MPI_Comm_rank(MPI_COMM_WORLD, &rank);
  MPI_Comm_size(MPI_COMM_WORLD, &nproc);
  seed_rng(rank);
  for (index=0;index<ITS;++index){
    /* random numbers between ±1*/
    d1=get_random_in_range();
    d2=get_random_in_range();
    if (d1*d1 + d2*d2 <= 1) count ++;
  }
  count_global = malloc(sizeof(int)*nproc);
  MPI_Gather(&count, 1, MPI_INT, count_global, 1, MPI_INT, 0, MPI_COMM_WORLD);
  if (rank == 0) {
    for (index = 0; index < nproc; ++index) {
      printf("%i,%i\n", count_global[index], ITS);
    }
  }
  MPI_Finalize();
}

File: Fortran/01_Starting/04_gather_parallel_dart.f90
MODULE support

  USE mpi
  IMPLICIT NONE

  !Use SELECTED_REAL_KIND to get a kind value for a double precision number
  INTEGER, PARAMETER :: dp = SELECTED_REAL_KIND(15, 307)
  INTEGER, PARAMETER :: its = 10000000

  CONTAINS

  SUBROUTINE seed_rng(rank)
    INTEGER, INTENT(IN) :: rank
    INTEGER :: rand_size, time
    INTEGER, DIMENSION(:), ALLOCATABLE :: seed

    !Fortran RNG seeding is a bit strange. This works
    CALL RANDOM_SEED(size=rand_size)
    ALLOCATE(seed(1:rand_size))
    CALL SYSTEM_CLOCK(time)
    seed = time + rank !Add rank to make random across processors
    CALL RANDOM_SEED(put=seed)
    DEALLOCATE(seed)
  END SUBROUTINE seed_rng



  FUNCTION get_random_in_range() RESULT(rnum)

    REAL(dp) :: rnum

    CALL RANDOM_NUMBER(rnum)
    rnum = rnum * 2.0_dp - 1.0_dp
  END FUNCTION get_random_in_range

END MODULE support

PROGRAM main
  USE support

  INTEGER :: count = 0, index, ierr, rank, nproc
  INTEGER, DIMENSION(:), ALLOCATABLE :: count_global
  REAL(dp) :: d1, d2

  CALL MPI_INIT(ierr)
  CALL MPI_COMM_SIZE(MPI_COMM_WORLD, nproc, ierr)
  CALL MPI_COMM_RANK(MPI_COMM_WORLD, rank, ierr)

  CALL seed_rng(rank)
  DO index = 0, its - 1
    ! random numbers between ±1
    d1=get_random_in_range()
    d2=get_random_in_range()
    if (d1*d1 + d2*d2 <= 1) count = count + 1
  END DO

  ALLOCATE(count_global(nproc))
  CALL MPI_GATHER(count, 1, MPI_INTEGER, count_global, 1, MPI_INTEGER, &
      0, MPI_COMM_WOLD, ierr)
  IF (rank == 0) THEN
    DO irank = 1, nproc
      PRINT *, count_global(irank), its
    END DO
  END IF
  DEALLOCATE(count_global)
  CALL MPI_FINALIZE(ierr)
END PROGRAM main
